### PR TITLE
Add "2ha.sh" script, managing 2-node Canonical K8s HA AA clusters

### PR DESCRIPTION
**Pull request by petrutlucian94**
Scenario overview:

* Canonical K8s cluster containing 2 nodes
* Dqlite data store (unable to obtain quorum)
* Primary node dqlite files stored on DRBD
  * sync block-level replication between the two nodes
* cluster monitoring and failover handled through Pacemaker

Script functionality:

* boostrap the service
  * wait for a DRBD primary to be elected
  * detect the node role based on the DRBD status and Dqlite state
    * have the replica wait for the primary to be ready before continuing
  * recover Dqlite after failovers
  * transfer and apply recovery files to secondary nodes
  * transfer Dqlite files to DRBD and other backup locations, creating necessary symlinks
* install required packages
* purge all K8s data
* clear Pacemaker taints
* remove recovery data

"2ha.sh start_service" is intended to be used as part of a systemd unit that bootstraps the k8s services, coordinating with the other node and taking any necessary steps to recover Dqlite.

This PR also adds a "how-to" guide for the 2-node A-A HA setup.

### CODE REVIEW

File: docs/src/snap/howto/2-node-ha.md [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+  passwordless sudo configured. For simplicity, the default "ubuntu" user can
+  be used.
+* We recommend using static IP configuration.
+
+The [2ha.sh script] automates most operations related to the 2-node HA scenario.
+Retrieve it like so:
+
+```
+sudo mkdir -p /var/snap/k8s/common
+repo=https://raw.githubusercontent.com/petrutlucian94/k8s-snap
```
Comment (bschimke95) [inline docs/src/snap/howto/2-node-ha.md:57]: This script should be part of the snap. If you put it into hack then it should either be available at `/snap/k8s/current/k8s/hack` or you would need to update the `snapcraft.yaml` to add this directory to the snap.

File: docs/src/snap/howto/2-node-ha.md [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+[Install]
+WantedBy=multi-user.target
+EOF
+
+sudo systemctl enable rc-local.service
+sudo systemctl start rc-local.service
+```
+
+Let's configure the DRBD block device that will hold the Dqlite data.
+Make sure to use the right node addresses.
```
Comment (bschimke95) [inline docs/src/snap/howto/2-node-ha.md:115]: Ensure the correct node addresses are used.

File: docs/src/snap/howto/2-node-ha.md [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+  }
+}
+EOF
+
+sudo drbdadm create-md r0
+sudo drbdadm status
+```
+
+Let's create a mount point for the DRBD block device. Non-default mount points
+need to be passed to the 2ha.sh script mentioned above, see the script for the
```
Comment (bschimke95) [inline docs/src/snap/howto/2-node-ha.md:146]: put `2ha.sh` into ticks `

File: docs/src/snap/howto/2-node-ha.md [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+full list of configurable parameters.
+
+```
+DRBD_MOUNT_DIR=/mnt/drbd0
+sudo mkdir -p $DRBD_MOUNT_DIR
+```
+
+Run the following *once* to initialize the filesystem.
```
Comment (bschimke95) [inline docs/src/snap/howto/2-node-ha.md:154]: Maybe add what would happen if you accidentally execute it multiple times?
e.g. "Executing this multiple times would cause ..."

File: docs/src/snap/howto/2-node-ha.md [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+
+Follow the above steps on both nodes before moving forward.
+
+### Pacemaker
+
+Let's define a Pacemaker resource for the DRBD block device, which
+ensures that the block device will be mounted on the replica in case of a
+primary node failure.
+
+Pacemaker fencing (stonith) configuration is environment specific and thus
```
Comment (bschimke95) [inline docs/src/snap/howto/2-node-ha.md:234]: nit: link to fencing/stonith explanation?

File: docs/src/snap/howto/2-node-ha.md [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+commit
+show
+quit
+EOF
+```
+
+Before moving forward, let's ensure that the DRBD Pacemaker resource runs on
+the primary (voter) Dqlite node.
+
+Remember that in our case only the primary node contains the latest Dqlite data,
```
Comment (bschimke95) [inline docs/src/snap/howto/2-node-ha.md:260]: In this setup, only the primary node holds the latest Dqlite data, which will be transferred to the DRBD device

File: docs/src/snap/howto/2-node-ha.md [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+sudo crm resource move fs_res <primary_node_name>
+
+# remove the node constraint.
+sudo crm resource clear fs_res
+```
+
+### Kubernetes services
+
+We can now turn our attention to the Kubernetes services. Ensure that the k8s
+snap services no longer start automatically. Instead, they will be manged by a
```
Comment (bschimke95) [inline docs/src/snap/howto/2-node-ha.md:283]: managed

File: docs/src/snap/howto/2-node-ha.md [hunk 8]
```diff
@@ -0,0 +1,423 @@
+# 2-Node Active-Active HA using Dqlite
+
+## Rationale
+
+High availability is a mandatory requirement for most production-grade
```
Comment (bschimke95) [inline docs/src/snap/howto/2-node-ha.md:5]: (HA)

File: docs/src/snap/howto/2-node-ha.md [hunk 9]
```diff
@@ -0,0 +1,423 @@
+# 2-Node Active-Active HA using Dqlite
```
Comment (bschimke95) [inline docs/src/snap/howto/2-node-ha.md:1]: High-Availability

File: k8s/hack/2ha.sh [hunk 10]
```diff
@@ -0,0 +1,823 @@
+#!/bin/bash
+
```
Comment (bschimke95) [inline k8s/hack/2ha.sh:2]: please add a description what this script does.

**Review by bschimke95 — COMMENTED** [review]
Did an initial pass. Consider the rephrasing as suggestions and feel free to ignore them

The script looks mostly fine (you already know my opinion on large bash scripts. Fine for now but should eventually be moved to Python or Go IMHO)

File: docs/src/snap/howto/2-node-ha.md (hunk 4 above)
Comment (petrutlucian94) [reply docs/src/snap/howto/2-node-ha.md:154]: Not much since we haven't yet stored any data yet, not sure if it's worth mentioning it.

The main reason why I emphasised "once" is that below there's a line saying ``Follow the above steps on both nodes before moving forward.``, however this step in particular doesn't have to be executed twice.

**Comment by petrutlucian94** [comment]
Thanks for reviewing this PR! I'll address the comments right away.

> The script looks mostly fine (you already know my opinion on large bash scripts. Fine for now but should eventually be moved to Python or Go IMHO)

I admit that [Openstack Devstack](https://github.com/openstack/devstack/tree/master/lib) changed my perception of what a large bash script means but I see your point.

File: docs/src/snap/howto/2-node-ha.md (hunk 1 above)
Comment (petrutlucian94) [reply docs/src/snap/howto/2-node-ha.md:57]: Great, I guess we can just use ``/snap/k8s/current/k8s/hack/2ha.sh``.

**Review by bschimke95 — APPROVED** [review]
LGTM

File: docs/src/snap/howto/2-node-ha.md [hunk 11]
```diff
@@ -0,0 +1,417 @@
...
+```
+sudo drbdadm up r0
+
+sudo drbdadm -- --overwrite-data-of-peer primary r0/0
+sudo mkfs.ext4 /dev/drbd0
+
+sudo drbdadm down r0
+```
+
+Add the drbd device to the ``multipathd`` blacklist, ensuring that the multipath
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:156]:
```suggestion
Add the DRBD device to the ``multipathd`` blacklist, ensuring that the multipath
```
Correct me if I am wrong here, I believe this would be capitalized as we have done so in other places.

File: docs/src/snap/howto/2-node-ha.md [hunk 12]
```diff
@@ -0,0 +1,417 @@
...
+}
+EOF
+
+sudo systemctl restart multipathd
+```
+
+### Corosync
+
+Let's prepare the Corosync configuration. Again, make sure to use the correct
+addresses.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:172]:
```suggestion
Prepare the Corosync configuration.
```

File: docs/src/snap/howto/2-node-ha.md [hunk 13]
```diff
@@ -0,0 +1,417 @@
...
+[Install]
+WantedBy=multi-user.target
+EOF
+
+sudo systemctl enable rc-local.service
+sudo systemctl start rc-local.service
+```
+
+Let's configure the DRBD block device that will hold the Dqlite data.
+Ensure the correct node addresses are used.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:106]:
```suggestion
Ensure the correct node addresses are used.
```
This is important, instead of repeating it later make it stand out by putting it in a note box. Example for a note: https://github.com/canonical/k8s-snap/blob/c74b9f580f92b2fc6b77bf4ad692eb348b62a17f/docs/src/charm/howto/charm.md?plain=1#L18-L21

File: docs/src/snap/howto/2-node-ha.md [hunk 14]
```diff
@@ -0,0 +1,417 @@
...
+The first step is to install the required packages:
+
+```
+/snap/k8s/current/k8s/hack/2ha.sh install_packages
+```
+
+### DRBD
+
+For the purpose of this guide, we are going to use a loopback device as DRBD
+backing storage:
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:63]:
```suggestion
This example uses a loopback device as DRBD
backing storage:
```

File: docs/src/snap/howto/2-node-ha.md [hunk 15]
```diff
@@ -0,0 +1,417 @@
...
+
+### Pacemaker
+
+Let's define a Pacemaker resource for the DRBD block device, which
+ensures that the block device will be mounted on the replica in case of a
+primary node failure.
+
+[Pacemaker fencing] (stonith) configuration is environment specific and thus
+outside the scope of this guide. However, we highly recommend using fencing
+if possible to reduce the risk of cluster split-brain situations.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:227]:
```suggestion
outside the scope of this guide. Using fencing is highly recommended
if it is possible to reduce the risk of cluster split-brain situations.
```

File: docs/src/snap/howto/2-node-ha.md [hunk 16]
```diff
@@ -0,0 +1,417 @@
...
+# remove the node constraint.
+sudo crm resource clear fs_res
+```
+
+### Kubernetes services
+
+We can now turn our attention to the Kubernetes services. Ensure that the k8s
+snap services no longer start automatically. Instead, they will be managed by a
+wrapper service.
+
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:276]:
```suggestion
### Managing Kubernetes Snap Services

For the two-node HA setup k8s snap services should no longer start automatically. Instead, they will be managed by a wrapper service.

```

File: docs/src/snap/howto/2-node-ha.md [hunk 17]
```diff
@@ -0,0 +1,417 @@
...
+```
+for f in `sudo snap services k8s  | awk 'NR>1 {print $1}'`; do
+    echo "disabling snap.$f"
+    sudo systemctl disable "snap.$f";
+done
+```
+
+The next step is to define the wrapper service. Add the following to
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:284]: You can consider making this a sub-header.

File: docs/src/snap/howto/2-node-ha.md [hunk 18]
```diff
@@ -0,0 +1,417 @@
...
+``/etc/systemd/system/2ha_k8s.service``. Note that the sample uses the
+``ubuntu`` user, feel free to use a different one as long as the prerequisites
+are met.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:287]: Consider making this a note.

File: docs/src/snap/howto/2-node-ha.md [hunk 19]
```diff
@@ -0,0 +1,417 @@
...
+
+[Install]
+WantedBy=multi-user.target
+```
+
+```{note}
+The ``2ha.sh start_service`` command used by the service wrapper automatically
+detects the expected Dqlite role based on the DRBD state and takes the
+necessary steps to bootstrap the Dqlite state directories, synchronize with the
+peer node (if available) and recover the database.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:310]: Break up this sentence in two please.

File: docs/src/snap/howto/2-node-ha.md [hunk 20]
```diff
@@ -0,0 +1,417 @@
...
+```
+
+We need the ``2ha_k8s`` service to be restarted once a DRBD failover occurs.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:313]:
```suggestion
When a DRBD failover occurs the ``2ha_k8s`` service needs to be restarted.
```

File: docs/src/snap/howto/2-node-ha.md [hunk 21]
```diff
@@ -0,0 +1,417 @@
...
+For that, we are going to define a separate service that will be invoked by
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:314]:
```suggestion
To accomplish this, define a separate service that will be invoked by
```

File: docs/src/snap/howto/2-node-ha.md [hunk 22]
```diff
@@ -0,0 +1,417 @@
...
+primitive ha_k8s_failover_service systemd:2ha_k8s_failover op start interval=0 timeout=120 op stop interval=0 timeout=30
+order failover_after_fs mandatory: fs_res:start ha_k8s_failover_service:start
+colocation fs_failover_colo INFINITY: fs_res ha_k8s_failover_service
+commit
+show
+quit
+EOF
+```
+
+The setup is ready, start the HA k8s service on both nodes:
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:355]:
```suggestion
Once the setup is complete on both nodes, start the two-node HA k8s service on each node:
```

File: docs/src/snap/howto/2-node-ha.md [hunk 23]
```diff
@@ -0,0 +1,417 @@
...
+Notice that ``2ha_k8s_failover`` must not be configured to start automatically,
+but instead is going to be managed through Pacemaker.
+
+```
+sudo systemctl enable 2ha_k8s
+sudo systemctl daemon-reload
+```
+
+Make sure that both nodes have been configured using the above steps before
+moving forward.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:339]: Please number the steps and then rephrase this to something like please do the following steps on both nodes before continuing with the setup.

File: docs/src/snap/howto/2-node-ha.md [hunk 24]
```diff
@@ -0,0 +1,417 @@
...
+
+We can now define a new Pacemaker resource that will invoke the
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:341]:
```suggestion
Define a new Pacemaker resource that will invoke the
```

File: docs/src/snap/howto/2-node-ha.md [hunk 25]
```diff
@@ -0,0 +1,417 @@
...
+
+The setup is ready, start the HA k8s service on both nodes:
+
+```
+sudo systemctl start 2ha_k8s
+```
+
+## Troubleshooting
+
+### Dqlite recovery failing because of unexpected data segments
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:363]: I believe we need to capitalize more "important" words in our headers here and in other places.

Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:363]: Add some words in between two headers please.

File: docs/src/snap/howto/2-node-ha.md [hunk 26]
```diff
@@ -0,0 +1,417 @@
...
+closed segment 0000000000002369-0000000000002655 is past last snapshot
+snapshot-2-2048-642428, pre-recovery backup:
+/var/snap/k8s/common/recovery-k8s-dqlite-2024-09-05T082644Z-pre-recovery.tar.gz
+```
+
+Remove the offending segments and restart the ``2ha_k8s`` service.
+
+### DRBD split brain
+
+The DRBD cluster may enter a split brain state and stop synchronizing. The
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/2-node-ha.md:379]: Please add a reference to a split brain explanation.

**Review by louiseschmidtgen — CHANGES_REQUESTED** [review]
Amazing work on the 2-node HA set-up @petrutlucian94!
Please iterate over my polishing comments. 
I am requesting changes because I would like to discuss the alternative solution with PostgreSQL.

File: docs/src/snap/howto/2-node-ha.md (hunk 24 above)
Comment (petrutlucian94) [reply docs/src/snap/howto/2-node-ha.md:341]: Right, so PostgreSQL was mentioned as an alternative in Antonio's document and it seemed like a good idea to cover it here as well. To be honest, I didn't get to try it out but I assumed it would be possible based on the Canonical Kubernetes docs on external datastores: https://documentation.ubuntu.com/canonical-kubernetes/latest/snap/howto/external-datastore/

File: docs/src/snap/howto/2-node-ha.md [hunk 27]
```diff
@@ -0,0 +1,417 @@
...
+
+The cluster monitoring and failover process will be handled by Pacemaker and
+Corosync. After a node failure, the DRBD volume will be mounted on the standby
+node, allowing access to the latest Dqlite database version.
+
+Additional recovery steps are automated and invoked through Pacemaker.
+
+## Alternatives
+
+Another possible approach is to use PostgreSQL with Kine and logical
```
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply docs/src/snap/howto/2-node-ha.md:34]: Right, so PostgreSQL was mentioned as an alternative in Antonio's document and it seemed like a good idea to cover it here as well. To be honest, I didn't get to try it out but I assumed it would be possible based on the Canonical Kubernetes docs on external datastores: https://documentation.ubuntu.com/canonical-kubernetes/latest/snap/howto/external-datastore/

I can remove the "alternatives" section for now.

File: docs/src/snap/howto/2-node-ha.md [hunk 28]
```diff
@@ -0,0 +1,417 @@
...
+### Prerequisites
+
+* Ensure both nodes are part of the Kubernetes cluster.
+  See the [getting started] and [add/remove nodes] guides.
+* The user associated with the HA service has SSH access to the peer node and
+  passwordless sudo configured. For simplicity, the default "ubuntu" user can
+  be used.
+* We recommend using static IP configuration.
+
+The [2ha.sh script] automates most operations related to the 2-node HA scenario
```
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply docs/src/snap/howto/2-node-ha.md:51]: Sure, I'd go with ``two-node-ha.sh``, it's easier to type when calling it.

File: docs/src/snap/howto/2-node-ha.md (hunk 12 above)
Comment (petrutlucian94) [reply docs/src/snap/howto/2-node-ha.md:172]: I'd keep the reminder to update the addresses so that the readers don't copy/paste the following by mistake:

```
HAONE_ADDR=<firstNodeAddress>
HATWO_ADDR=<secondNodeAddress>
```

File: docs/src/snap/howto/2-node-ha.md (hunk 13 above)
Comment (petrutlucian94) [reply docs/src/snap/howto/2-node-ha.md:106]: Those are just some reminders to fill in the addresses instead of copy/pasting the following by mistake:

```
HAONE_ADDR=<firstNodeAddress>
HATWO_ADDR=<secondNodeAddress>
```

Do we need a note box for that?

File: docs/src/snap/howto/2-node-ha.md (hunk 25 above)
Comment (petrutlucian94) [reply docs/src/snap/howto/2-node-ha.md:363]: We could, although the other docs don't seem to do this:

* https://documentation.ubuntu.com/canonical-kubernetes/latest/snap/howto/external-datastore/

* https://documentation.ubuntu.com/canonical-kubernetes/latest/snap/howto/backup-restore/

Furthermore, it might affect readability.

File: docs/src/snap/howto/2-node-ha.md (hunk 23 above)
Comment (petrutlucian94) [reply docs/src/snap/howto/2-node-ha.md]: Should we number each step starting with ```### Prerequisites```?

**Comment by petrutlucian94** [comment]
@louiseschmidtgen Thanks for reviewing the docs! I've addressed most comments and left a few questions.

File: docs/src/snap/howto/two-node-ha.md [hunk 29]
```diff
@@ -0,0 +1,426 @@
...
+EOF
+
+sudo chmod +x /etc/rc.local
+```
+
+Add a service to automatically execute the ``/etc/rc.local`` script.
+
+```
+
+
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/two-node-ha.md:73]:
```suggestion
```
Please remove these empty spaces

File: docs/src/snap/howto/two-node-ha.md [hunk 30]
```diff
@@ -0,0 +1,426 @@
...
+}
+EOF
+
+sudo drbdadm create-md r0
+sudo drbdadm status
+```
+
+Create a mount point for the DRBD block device. Non-default mount points
+need to be passed to the ``two-node-ha.sh`` script mentioned above, see the
+script for the full list of configurable parameters.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/two-node-ha.md:130]:
```suggestion
need to be passed to the ``two-node-ha.sh`` script mentioned above. Please refer to the
script for the full list of configurable parameters.
```

File: docs/src/snap/howto/two-node-ha.md [hunk 31]
```diff
@@ -0,0 +1,426 @@
...
+
+```
+DRBD_MOUNT_DIR=/mnt/drbd0
+sudo mkdir -p $DRBD_MOUNT_DIR
+```
+
+Run the following *once* to initialize the filesystem.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/two-node-ha.md:137]: Does something bad happen if we run it twice that users should know about?

File: docs/src/snap/howto/two-node-ha.md [hunk 32]
```diff
@@ -0,0 +1,426 @@
...
+
+### Pacemaker
+
+Let's define a Pacemaker resource for the DRBD block device, which
+ensures that the block device will be mounted on the replica in case of a
+primary node failure.
+
+[Pacemaker fencing] (stonith) configuration is environment specific and thus
+outside the scope of this guide. Using fencing is highly recommended if it is
+possible to reduce the risk of cluster split-brain situations.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/two-node-ha.md:219]:
```suggestion
outside the scope of this guide. Using fencing is highly recommended, if it is
possible, to reduce the risk of cluster split-brain situations.
```
Commas here are optional but I believe this helps readability.

File: docs/src/snap/howto/2-node-ha.md [hunk 33]
```diff
@@ -0,0 +1,417 @@
+# 2-Node Active-Active High-Availability using Dqlite
```
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply docs/src/snap/howto/2-node-ha.md]: Whops, sure.

File: docs/src/snap/howto/two-node-ha.md [hunk 34]
```diff
@@ -0,0 +1,426 @@
...
+```
+sudo crm resource move fs_res <primary_node_name>
+
+# remove the node constraint.
+sudo crm resource clear fs_res
+```
+
+### Managing Kubernetes Snap Services
+
+For the two-node HA setup k8s snap services should no longer start
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/two-node-ha.md:265]:
```suggestion
For the two-node HA setup, k8s snap services should no longer start
```
This comma is not required either but helps the flow.

File: docs/src/snap/howto/two-node-ha.md [hunk 35]
```diff
@@ -0,0 +1,426 @@
...
+
+```{note}
+The ``two-node-ha.sh start_service`` command used by the service wrapper
+automatically detects the expected Dqlite role based on the DRBD state.
+It then takes the necessary steps to bootstrap the Dqlite state directories,
+synchronize with the peer node (if available) and recover the database.
+```
+
+When a DRBD failover occurs, the ``two-node-ha-k8s`` service needs to be
+restarted. To accomplish this,, we are going to define a separate service that
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/two-node-ha.md:310]:
```suggestion
restarted. To accomplish this, we are going to define a separate service that
```
Okay now this second comma we don't need :)

File: docs/src/snap/howto/two-node-ha.md [hunk 36]
```diff
@@ -0,0 +1,426 @@
...
+
+```
+Error: failed to recover k8s-dqlite, error: k8s-dqlite recovery failed, error:
+recover failed with error code 1, error details: raft_recover(): io:
+closed segment 0000000000002369-0000000000002655 is past last snapshot
+snapshot-2-2048-642428, pre-recovery backup:
+/var/snap/k8s/common/recovery-k8s-dqlite-2024-09-05T082644Z-pre-recovery.tar.gz
+```
+
+Remove the offending segments and restart the ``two-node-ha-k8s`` service.
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/two-node-ha.md:378]: As a user I would wonder if it is save to remove this segment. Should we add an explanation? This is the data that has not been replicated to more than half of the voters and can therefore be safely removed?

File: docs/src/snap/howto/two-node-ha.md [hunk 37]
```diff
@@ -0,0 +1,426 @@
...
+
+Follow the above steps on both nodes before moving forward.
+
+### Pacemaker
+
+Let's define a Pacemaker resource for the DRBD block device, which
+ensures that the block device will be mounted on the replica in case of a
+primary node failure.
+
+[Pacemaker fencing] (stonith) configuration is environment specific and thus
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/two-node-ha.md:217]:
```suggestion
[Pacemaker fencing] (stonith) configuration is environment specific and thus
```
We should at least write out stonith once on the first mention.

File: docs/src/snap/howto/two-node-ha.md [hunk 38]
```diff
@@ -0,0 +1,426 @@
...
+
+ubuntu@hatwo:~$ cat /proc/drbd
+version: 8.4.11 (api:1/proto:86-101)
+srcversion: C7B8F7076B8D6DB066D84D9
+ 0: cs:StandAlone ro:Secondary/Unknown ds:UpToDate/DUnknown   r-----
+    ns:0 nr:0 dw:0 dr:0 al:0 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:1802140
+
+ubuntu@hatwo:~$ dmesg | grep "Split"
+[  +0.000082] block drbd0: Split-Brain detected but unresolved, dropping connection!
+
```
Comment (louiseschmidtgen) [inline docs/src/snap/howto/two-node-ha.md:398]:
```suggestion
```
Remove the extra line please.

File: k8s/hack/two-node-ha.sh [hunk 39]
```diff
@@ -0,0 +1,826 @@
...
+            sleep $pollInterval
+        fi
+    done
+
+    log_message "Timed out waiting for DRBD resource." \
+                "Waited: ${SECONDS}. Timeout: ${DRBD_READY_TIMEOUT}s."
+    return 1
+}
+
+# Based on the drbd volume state, we decide if this node should be a
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:359]:
```suggestion
# Based on the DRBD volume state, we decide if this node should be a
```

File: docs/src/snap/howto/2-node-ha.md (hunk 33 above)
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply docs/src/snap/howto/2-node-ha.md]: I guess we can just say that ``Both nodes will be active members of the cluster, sharing the Kubernetes load.``

File: k8s/hack/two-node-ha.sh [hunk 40]
```diff
@@ -0,0 +1,826 @@
...
+        # Stop the k8s snap directly instead of the wrapper service so that
+        # we won't cause failures if both nodes start at the same time.
+        # The secondary will wait for the k8s services to start on the primary.
+        if ssh $SSH_OPTS $SSH_USERNAME@$peerIp sudo snap stop k8s; then
+            stoppedPeer=1
+            log_message "Successfully stopped peer k8s services."
+            log_message "The stopped services are going to be restarted after the recovery finishes."
+        else
+            log_message "Couldn't stop k8s services on the peer node." \
+                        "Assuming that it's stopped and proceeding with the recovery."
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:460]:
```suggestion
                        "Assuming that the k8s service is stopped and proceeding with the recovery."
```

File: k8s/hack/two-node-ha.sh [hunk 41]
```diff
@@ -0,0 +1,826 @@
...
+        # directly and the wrapper service doesn't currently monitor it.
+        ssh $SSH_OPTS $SSH_USERNAME@$peerIp sudo systemctl restart $SYSTEMD_SERVICE_NAME ||
+            log_message "Couldn't start peer k8s services."
+    fi
+}
+
+function process_recovery_files_on_secondary() {
+    local peerIp="$1"
+
+    log_message "Ensuring that the drbd volume is unmounted."
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:509]:
```suggestion
    log_message "Ensuring that the DRBD volume is unmounted."
```

File: k8s/hack/two-node-ha.sh [hunk 42]
```diff
@@ -0,0 +1,826 @@
...
+}
+
+function start_service() {
+    log_message "Initializing node."
+
+    # DRBD is the primary source of truth for the dqlite role.
+    # We need to wait for it to become available.
+    wait_drbd_resource
+
+    # dump the drbd and pacemaker status for debugging purposes.
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:676]:
```suggestion
    # dump the DRBD and pacemaker status for debugging purposes.
```

File: k8s/hack/two-node-ha.sh [hunk 43]
```diff
@@ -0,0 +1,826 @@
...
+
+function clean_recovery_data() {
+    log_message "Cleaning up dqlite recovery data."
+    rm -f $K8SD_RECOVERY_TARBALL
+    rm -f $K8SD_RECOVERY_TARBALL_BKP
+    rm -f $K8S_DQLITE_STATE_DIR/recovery-k8s-dqlite*
+}
+
+function purge() {
+    log_message "Removing the k8s snap and all the associated files."
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:726]:
```suggestion
    log_message "Removing the k8s-snap and all the associated files."
```

File: k8s/hack/two-node-ha.sh [hunk 44]
```diff
@@ -0,0 +1,826 @@
...
+Commands:
+    move_statedirs          Move the dqlite state directories to the DRBD mount,
+                            replacing them with symlinks.
+                            The existing contents are moved to a backup folder,
+                            which can be used as part of the recovery process.
+    install_packages        Install the packages required by the two-node HA
+                            cluster.
+    start_service           Initialize the k8s services, taking the following
+                            steps:
+                            1. Based on the drbd state, decide if this node
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:801]:
```suggestion
                            1. Based on the DRBD state, decide if this node
```

File: k8s/hack/two-node-ha.sh [hunk 45]
```diff
@@ -0,0 +1,826 @@
...
+                               should assume the primary (dqlite voter) or
+                               secondary (spare) role.
+                            2. If this is the first start, transfer the dqlite
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:804]:
```suggestion
                            2. If this is the first start, transfer the Dqlite
```

File: k8s/hack/two-node-ha.sh [hunk 46]
```diff
@@ -0,0 +1,826 @@
...
+                               state directories and create backups.
+                            3. If this node is a primary, promote it and initiate
+                               the dqlite recovery, creating recovery tarballs.
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:807]:
```suggestion
                               the Dqlite recovery, creating recovery tarballs.
```

File: k8s/hack/two-node-ha.sh [hunk 47]
```diff
@@ -0,0 +1,826 @@
+#!/bin/bash
+
+# This script automates various operations on two-node HA A-A Canonical K8s
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:3]:
```suggestion
# This script automates various operations on two-node HA Active-Active Canonical K8s
```

File: k8s/hack/two-node-ha.sh [hunk 48]
```diff
@@ -0,0 +1,826 @@
...
+# clusters that use the default datastore, Dqlite.
+#
+# Prerequisites:
+# * required packages installed using the "install_packages" command.
+# * initialized K8s cluster, both nodes joined
+# * the current user has ssh access to the peer node.
+#   - used to handle K8s services and transfer Dqlite data
+# * the current user has passwordless sudo enabled.
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:11]:
```suggestion
```
Remove duplicated comment

File: k8s/hack/two-node-ha.sh [hunk 49]
```diff
@@ -0,0 +1,826 @@
...
+
+function get_dqlite_node_role() {
+    local infoYamlPath=$1
+    sudo cat $infoYamlPath | yq -r '.Role'
+}
+
+function get_dqlite_role_from_cluster_yaml() {
+    # Note that the cluster.yaml role may not match the info.yaml role.
+    # In case of a freshly joined node, info.yaml will have "voter" role
+    # while cluster.yaml has "spare" role.
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:91]:
```suggestion
    # In case of a freshly joined node, info.yaml will show it as a "voter"
    # while cluster.yaml lists it as a "spare" role.
```

File: k8s/hack/two-node-ha.sh [hunk 50]
```diff
@@ -0,0 +1,826 @@
...
+
+    if [[ ${#addresses[@]} -lt 1 ]]; then
+        log_message "No dql peers found."
+        exit 1
+    fi
+
+    echo ${addresses[0]} | cut -d ":" -f 1
+}
+
+# This function moves the dqlite state directories to the DRBD mount,
```
Comment (louiseschmidtgen) [inline k8s/hack/two-node-ha.sh:149]:
```suggestion
# This function moves the Dqlite state directories to the DRBD mount,
```

**Review by louiseschmidtgen — APPROVED** [review]
Couple more small comment that need to be addressed, afterwards we are good to go. Thank you @petrutlucian94

File: docs/src/snap/howto/two-node-ha.md (hunk 31 above)
Comment (petrutlucian94) [reply docs/src/snap/howto/two-node-ha.md:137]: Not really. It's just that unlike the other commands,  this doesn't have to be called twice.

File: docs/src/snap/howto/two-node-ha.md (hunk 36 above)
Comment (petrutlucian94) [reply docs/src/snap/howto/two-node-ha.md:378]: In our scenario, Dqlite doesn't replicate data at all, which is why we need DRBD. Removing data segments could lead to data loss, but there's no other documented way of recovering the cluster.

To be honest, the dqlite documentation is rather scarce and doesn't even mention this. I had to rely on the LXD documentation as well as the comments on GitHub issues:

* https://documentation.ubuntu.com/lxd/en/latest/howto/dqlite_troubleshoot/#interventions
* https://github.com/canonical/dqlite/issues/309#issuecomment-854736238

File: k8s/hack/two-node-ha.sh (hunk 43 above)
Comment (petrutlucian94) [reply k8s/hack/two-node-ha.sh:726]: ``k8s`` is the name of the snap, not sure if the dash is appropriate here.

File: k8s/hack/two-node-ha.sh (hunk 48 above)
Comment (petrutlucian94) [reply k8s/hack/two-node-ha.sh:11]: whops

File: docs/src/snap/howto/two-node-ha.md (hunk 36 above)
Comment (louiseschmidtgen) [reply docs/src/snap/howto/two-node-ha.md:379]: I see, can we add a warning similar to the one in LXD's docs please?

**Comment by bschimke95** [comment]
Great work @petrutlucian94!
